Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp

```diff
@@ -145,15 +146,117 @@ s32 NWC24MakeUserID(u64* nwc24_id, u32 hollywood_id, u16 id_ctr, HardwareModel h
 }  // Anonymous namespace
 
 NetKDRequestDevice::NetKDRequestDevice(Kernel& ios, const std::string& device_name)
-    : Device(ios, device_name), config{ios.GetFS()}
+    : Device(ios, device_name), config{ios.GetFS()}, dlList{ios.GetFS()}
 {
+  m_work_queue.Reset([this](AsyncTask task) {
+    const IPCReply reply = task.handler();
+    {
+      std::lock_guard lg(m_async_reply_lock);
+      m_async_replies.emplace(AsyncReply{task.request, reply.return_value});
+    }
+  });
 }
 
 NetKDRequestDevice::~NetKDRequestDevice()
 {
   WiiSockMan::GetInstance().Clean();
 }
 
+void NetKDRequestDevice::Update()
+{
+  {
+    std::lock_guard lg(m_async_reply_lock);
+    while (!m_async_replies.empty())
+    {
+      const auto& reply = m_async_replies.front();
+      GetIOS()->EnqueueIPCReply(reply.request, reply.return_value);
+      m_async_replies.pop();
+    }
+  }
+}
+
+IPCReply NetKDRequestDevice::HandleNWC24DownloadNowEx(const IOCtlRequest& request)
+{
+  std::vector<u8> file_data;
+
+  u32 flags = Memory::Read_U32(request.buffer_in);
+  // Nintendo converts the entry ID between a u32 and u16
+  // several times, presumably for alignment purposes.
+  // We'll skip past buffer_in+4 and keep the entry index as a u16.
+  u16 entryIndex = Memory::Read_U16(request.buffer_in + 6);
+  u32 subtaskBitmask = Memory::Read_U32(request.buffer_in + 8);
+
+  INFO_LOG_FMT(IOS_WC24,
+               "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - flags: {}, index: {}, bitmask: {}",
+               flags, entryIndex, subtaskBitmask);
+
+  INFO_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - URL: {}",
+               dlList.GetDownloadURL(entryIndex));
+
+  INFO_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - Name: {}",
+               dlList.GetVFFContentName(entryIndex));
+
+  INFO_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - Subtask: {}",
+               dlList.SubtaskID(entryIndex));
+
+  const Common::HttpRequest::Response response = m_http.Get(dlList.GetDownloadURL(entryIndex));
+
+  if (!response)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to request data at {}", dlList.GetDownloadURL(entryIndex));
+    WriteReturnValue(NWC24::WC24_ERR_NETWORK, request.buffer_out);
+    return IPCReply(NWC24::WC24_ERR_NETWORK);
+  }
+
+  // Write buffer to temporary file
+  const auto temp = m_ios.GetFS()->CreateAndOpenFile(
+      PID_KD, PID_KD, DL_CNT_PATH, {FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::ReadWrite});
+  if (!temp)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to create dlcnt.bin");
+    WriteReturnValue(NWC24::WC24_ERR_FILE_OPEN, request.buffer_out);
+    return IPCReply(NWC24::WC24_ERR_FILE_OPEN);
+  }
+
+  temp->Write(response->data(), response->size());
+
+  // Now we read the file
+  NWC24::WC24File wc24File;
+  temp->Seek(0, FS::SeekMode::Set);
+  temp->Read(&wc24File, 1);
+
+  std::vector<u8> tempBuffer = std::vector<u8>(response->size() - 320);
```

Comment:
I think, the following should be equivalent:
```C++
std::vector<u8> temp_buffer(response->size() - 320);
```

Is `response->size()` guaranteed to be always `>= 320`?

Reply:
Yes, all WiiConnect24 downloaded content is either in the WC24 container that has a header size of 320 bytes, or just 320 null bytes before the actual file data.

Reply:
As far as I can tell this comes from the network, so you should verify that the data is actually >= 320 bytes. Else you can blow up Dolphin with a malicious WC24 server.

Reply:
Sounds good, also should we consider verifying the RSA signature? I know this may conflict with RiiConnect24 because I believe they patched IOS 80 to use another certificate.

Reply:
How does something like this look?

```cpp
if (response->size() >= 320) {
    ERROR_LOG_FMT(IOS_WC24, "File at URL {} contained invalid data", url);
    WriteReturnValue(NWC24::WC24_ERR_FILE_BROKEN, request.buffer_out);
    return IPCReply(NWC24::WC24_ERR_FILE_BROKEN);
}
```

`WC24_ERR_FILE_BROKEN` is -46

Again, I would like to maybe look into verifying the RSA signature as another precaution.

Reply:
Not sure if that's going to be useful - the original servers don't exist anymore anyways, so anything this could download would use a custom, non-Nintendo signature. That would mean users would have to manually take RC24's public key and add it to the Dolphin config? Or the RC24 public key would be built into Dolphin, not sure if that's a good idea. Or am I missing something?